### DISCUSSION TITLE
User-defined functions for thresholds in gen_ens_prod

### DISCUSSION THREAD

**Discussion by Chris-J-Steele** [dtcenter/METplus#1287] [Configuration] [answered]
Hello,

In the MET documentation chapter describing the configuration file overview (https://met.readthedocs.io/en/develop/Users_Guide/config_options.html), it talks about acceptable input types. One of these is a user defined function. The description of the required formatting is described below:

      -Left side is a function name followed by variable name in parenthesis.

      -Right side is an equation which includes basic math functions (+,-,*,/), built-in functions (listed below), or other user-defined 
      functions.

      -Built-in functions include: sin, cos, tan, sind, cosd, tand, asin, acos, atan, asind, acosd, atand, atan2, atan2d, arg, argd, log, exp, 
      log10, exp10, sqrt, abs, min, max, mod, floor, ceil, step, nint, sign   

Is there an example of this anywhere that I can look at? What I would like to do is to define my ensemble field thresholds based on the mean climatology +-  multiples of the climatological standard deviation. Is this currently possible with MET?

I can see in the documentation that it needs the function to be defined using a single variable name. Does this have to be in the forecast file or can it be from the climatology files?

regards

Chris

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hi Chris, I see you are looking for examples of defining functions in MET configuration files. MET's ConfigConstants file includes some function examples for common unit conversions. Take a look starting at this line:
https://github.com/dtcenter/MET/blob/287cd5d1f5fe55f803c95c2b620d106a76fedd65/met/data/config/ConfigConstants#L28

However, I believe that functions are currently only used in the MET configuration files in 2 spots:
- To define the "convert" configuration option discussed in this [section](https://met.readthedocs.io/en/develop/Users_Guide/config_options.html?highlight=convert#fcst) of the User's Guide.
- To define the interest functions for the fuzzy logic in [MODE](https://github.com/dtcenter/MET/blob/287cd5d1f5fe55f803c95c2b620d106a76fedd65/met/data/config/MODEConfig_default#L128) and [MODE Time-Domain (mtd)](https://github.com/dtcenter/MET/blob/287cd5d1f5fe55f803c95c2b620d106a76fedd65/met/data/config/MTDConfig_default#L120).

The convert function is defined as a function of 1 variable. But actually, the interest functions are piecewise linear functions... that's why their syntax looks different.

Ultimately though you'd like define thresholds relative to a climatological distribution. Please take a look in this [section](https://met.readthedocs.io/en/develop/Users_Guide/config_options.html) of the User's Guide and search for "Percentile Thresholds". Specifically, please read about Climatological Distribution Percentile (CDP) thresholds. Assuming you've provided a climatological mean and standard deviation fields (using climo_mean and climo_stdev config options), you can define thresholds relative to that distribution. MET currently assumes a normal distribution defined by the mean and stdev at each grid point. I believe that adding/subtracting a certain number of standard deviations is possible by just selecting the right percentile. For example, ">CDP68" should be the same as "greater than the mean + 1 standard deviation".

Hope that helps clarify.

**Reply by JohnHalleyGotway** [reply]
@Chris-J-Steele I just want to followup to see if you have any remaining questions. If not, you could select the "**Mark as answer**" button for my response. Once a discussion has been marked as answered, we then "Lock conversation" on it to prevent future posts. That encourages users to create new discussions for new questions, and also makes it easier for us to keep track of open questions.

**Reply by Chris-J-Steele** [reply]
Sorry @JohnHalleyGotway for the delayed response. Thanks for the answer, that does make sense, however I am a little concerned about the assumption that the distribution is normal. Is there any possibility that MET will allow for an empirical standard deviation like it does in the CRPS now?

**Reply by JohnHalleyGotway** [reply]
Chris, I understand your concern about the assumption of normality. Currently, there is no mechanism within MET to use a full empirical distribution of values. There would be several challenges in implementing that, the largest of which is efficiency. Sifting through a full empirical distribution of values for each grid point would increase the processing requirements by an order of magnitude if that were done on the fly within each MET tool run.

I can think of a few options that might be more realistic.

1. Users can always define their own percentile threshold upstream of calling the MET tools. For example, perhaps you process a full season's worth of data and determine that a threshold of 2.65 corresponds to the 75-th percentile of the data. You can set the MET config file to use ">USP75(2.65)". It'll use the ">2.65" threshold and write ">USP75(2.65)" to the output files. "USP" stands for user-specified-threshold. But this isn't all that helpful in this case. That threshold value is applied to all grid points... meaning that it doesn't vary based on the climatology at that point.

2. There's nothing particularly special about assuming a normal distribution. It would be pretty easy to enhance MET to support different types of distributions. In fact, it already does this for the OBS_ERROR_TABLE used by Ensemble-Stat and described in this [chapter](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=obs_error#configuration-file-overview). I suspect the difficult part would be choosing (and justifying) which distributions are more appropriate for which variables.

3. Another option would be a variation of (1) listed above. We could consider letting the user define a gridded field of data which represents the 75-th percentile of data computed over some time period. Then define the threshold relative to the value of that field. That would enable the user-specified-percentile value change from grid point to grid point.

So no easy, obvious solutions but some ideas to consider.

**Reply by Chris-J-Steele** [reply]
OK, thanks very much for explaining. I'll have a think about which is the best method for us.
